api/src/main/java/net/kyori/adventure/text/width/PixelWidthSource.java

```diff
@@ -0,0 +1,138 @@
+   * @return a pixel width source using a {@link PixelWidthSourceImpl#DEFAULT_FONT_WIDTH character function} for the default minecraft font
+   * @since 4.5.0
+   */
+  static <CX> @NonNull PixelWidthSource<CX> defaultPixelWidth(final @NonNull Function<CX, Locale> localeFunction){
+    return new PixelWidthSourceImpl<>(cx -> PixelWidthSourceImpl.DEFAULT_FONT_WIDTH, localeFunction);
+  }
+
+  /**
+   * A pixel width source calculating width using the default minecraft font.
+   *
+   * @param <CX> the context for this pixel width source(e.g player, world, locale)
+   * @param localeFunction used for automatic server-side translations using the {@link GlobalTranslator}
+   * @return a pixel width source using a custom character function
+   * @since 4.5.0
+   */
+  static <CX> @NonNull PixelWidthSource<CX> withCustomCharacterFunction(final @NonNull Function<CX, CharToIntFunction<Style>> function, final @NonNull Function<CX, Locale> localeFunction){
+    return new PixelWidthSourceImpl<>(function, localeFunction);
+  }
+
+  /**
+   * Calculates the pixel width of a component, given a context.
+   *
+   * @param component a component
+   * @param context the context of this calculation
+   * @return the pixel width of the component
+   * @since 4.5.0
+   */
+  int width(final @NonNull Component component, final @NonNull CX context);
+
+  /**
+   * Calculates the pixel width of a string, given a context.
+   *
+   * @param string a string
+   * @param style the style of the string
+   * @param context the context of this calculation
+   * @return the pixel width of the string
+   * @since 4.5.0
+   */
+  int width(final @NonNull String string, final @NonNull Style style, final @NonNull CX context);
+
+  /**
+   * Calculates the pixel width of a character, given a context.
+   *
+   * @param c a character
+   * @param style the style of the character
+   * @param context the context of this calculation
+   * @return the pixel width of the character
+   * @since 4.5.0
+   */
+  int width(final char c, final @NonNull Style style, final @NonNull CX context);
+
+  /**
+   * Add a {@link ComponentResolver} to this source that can resolve a specific type of component given a context.
+   * Whenever a component is resolved successfully the result is used regardless of if other resolvers are able to resolve as well.
+   *
+   * @param resolveFor the component type this resolver accepts
+   * @param resolver the resolver
+   * @see ComponentResolver
+   * @since 4.5.0
+   */
+  <CO extends Component> void addResolver(final @NonNull Class<CO> resolveFor, final @NonNull ComponentResolver<CO, CX> resolver);
+
+  /**
+   * Something that can either resolve a {@link Component} into a {@link TextComponent}, or return {@code null} if not possible.
+   *
+   * @param <CX> a context (player, server, locale)
+   * @since 4.5.0
+   */
+  interface ComponentResolver<CO extends Component, CX> {
```

Comment:
I think this component resolver system could be broken out to be more broadly applicable for operations that need to work with linear text -- think the legacy (#135) and plain serializers, or selector and NBT components that need server-side resolving (KyoriPowered/adventure-platform#8)